server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -881,6 +881,7 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
         long startTime = System.nanoTime();
         assert source.dynamicTemplates().isEmpty() || origin == Engine.Operation.Origin.PRIMARY :
             "dynamic_templates parameter can only be associated with primary operations";
+        mapperService.validateType(type);
```

Comment:
Another naive question (sorry): could we catch this in `MapperService#merge` rather than relying on callers checking they're not adding a type first? Not to say there's anything wrong with checking it here, just that there's a risk there's more code paths that aren't covered. Maybe not today but possibly in a future version.

Reply:
Types are gone entirely in 8x so I'm not that concerned about future new code paths.  I did initially try detecting this in `MapperService#merge` but that ended up breaking loads of unit tests which aren't being careful with their types, so I felt that this was the best fix for the minimum footprint.

Reply:
Ok, makes sense, I also expect this won't change too much in what's left of the 7.x series. I would like this to be reviewed by someone who knows more about mappings than me, but this LGTM.